csi.proto

```diff
@@ -194,6 +194,18 @@ message CreateVolumeRequest {
   // sensitive and MUST be treated as such (not logged, etc.) by the CO.
   // This field is OPTIONAL.
   map<string, string> controller_create_credentials = 6;
+
+  // The topological segment(s) the provisioned volume MUST be
+  // accessible from.
+  // A topological segment is a sub-division of a cluster, like region,
+  // zone, rack, etc.
+  // An SP SHALL advertise the requirements for topology in
```

Comment:
Is it possible for the user to configure sites/racks/zones and regions in kubernetes that an SP can query.  So that both the CO and SP have the same view for what topology elements are available. What would the definition include and how does the SP figure that it can access and provision volumes at a zone/region/rack etc.?

Reply:
> How does the CO (or for that matter an SP) enumerate a region/zone or rack in the topology? Does the SP need to query and pull this data from the CO?

If the volumes that an SP makes available have availability constraints the SP should have some mechanism to enumerate those topological segments. The SP communicates those to the CO via the `NodeGetId` call. So, the CO "enumerates" the SP's topology via the `NodeGetId` call (i.e. retrieve the node/zone/rack/etc., as defined by SP, that each node belongs to).

> HOw are changes in the topology or access to a region or rack notified either way? For example, CO (or SP) detects a change in toplogy or looses access to a rack/zone etc.

The topology for a node is retrieved via a `NodeGetId` call when the plugin is registered on the node. The topology for a volume is returned in the `CreateVolume` response. Once a volume or node are provisioned there topology is not expected to change. In the odd chance that it does, reinitializing those elements will allow picking up the new topology info.

> Is it possible for the user to configure sites/racks/zones and regions in kubernetes that an SP can query. So that both the CO and SP have the same view for what topology elements are available. 

In the proposed design, when a new plugin is registered, Kubernetes will use `NodeGetId` to retrieve the SP's topology information for each node and apply it as labels to the Kubernetes node object. These labels can then be used with all the familiar Kubernetes mechanisms (PodAffinity, etc.).

Making SP aware of any CO specific topology information is not part of this design, because it is unnecessary.